limacharlie/firehose.go

```diff
@@ -0,0 +1,308 @@
+	if err != nil {
+		return fmt.Errorf("could not register output with name '%s': %s", outputName, err)
+	}
+
+	_, found := allOutputs[outputName]
+	if found {
+		log.Debug().Str("name", outputName).Msg("output registration already done")
+		return nil
+	}
+
+	output := makeGenericOutput(outputOpts)
+	_, err = org.OutputAdd(output)
+	if err != nil {
+		return fmt.Errorf("could not add output: %s", err)
+	}
+	log.Debug().Msg("output registration done")
+	return nil
+}
+
+func StartFirehose(org Organization, fhOpts FirehoseOptions) (Firehose, error) {
+	return StartFirehoseAndRegisterOutput(org, fhOpts, nil)
+}
+
+func StartFirehoseAndRegisterOutput(org Organization, fhOpts FirehoseOptions, fhOutputOpts *FirehoseOutputOptions) (Firehose, error) {
+	if fhOutputOpts != nil {
+		err := registerOutput(org, *fhOutputOpts)
+		if err != nil {
+			return Firehose{}, err
+		}
+	}
+
+	listener, err := startListener(fhOpts.ListenOnIP, fhOpts.ListenOnPort, fhOpts.SSLCertPath, fhOpts.SSLCertKeyPath)
+	if err != nil {
+		return Firehose{}, err
+	}
+
+	fh := Firehose{org,
+		fhOpts,
+		fhOutputOpts,
+		make(chan FirehoseMessage, fhOpts.MaxMessageCount),
+		make(chan FirehoseMessage, fhOpts.MaxErrorMessageCount),
+		0,
+		*listener,
+		make(chan bool, 1)}
+	go fh.handleConnections()
+	return fh, nil
+}
+
+func (fh Firehose) handleConnections() {
+	readBufferSize := 1024 * 512
+	currentData := make([]byte, readBufferSize*2)
+
+	log.Debug().Msg("listening for connections")
+	for fh.IsRunning() {
+		conn, err := fh.listener.Accept()
+		if err != nil {
+			continue
+		}
+		log.Debug().Msg("new incoming connection")
+		defer conn.Close()
+
+		readBuffer := make([]byte, readBufferSize)
+		sizeRead, err := conn.Read(readBuffer)
+		if err != nil {
+			log.Err(err).Msg("error reading from connection")
+			continue
+		}
+		if sizeRead == 0 {
+			log.Debug().Msg("empty body read")
+			continue
+		}
+		chunks := bytes.Split(readBuffer, []byte{0x0a})
+		isContinuation := len(chunks) == 1
+		if isContinuation {
+			currentData = append(currentData[:len(currentData)], chunks[0]...)
+			continue
+		}
+		for _, chunk := range chunks {
+			currentData = append(currentData[:len(currentData)], chunk...)
+			if len(currentData) == 0 {
+				continue
+			}
+			fh.handleMessage(currentData)
+			currentData = append(currentData[:0], currentData[len(currentData):]...)
+		}
+	}
+}
+
+func (fh Firehose) handleMessage(raw []byte) {
+	fhMessage := FirehoseMessage{string(raw)}
+	if fh.opts.ParseMessage {
+		isValid := json.Valid(raw)
+		if isValid && len(fh.Messages) < fh.opts.MaxMessageCount {
+			fh.Messages <- fhMessage
+		} else {
+			fh.messageDropCount++
+			if len(fh.ErrorMessages) < fh.opts.MaxErrorMessageCount {
+				fh.ErrorMessages <- fhMessage
+			} else {
+				log.Warn().Msg("maximum error message count reached")
+			}
+		}
+	} else {
+		fh.Messages <- fhMessage
+	}
+}
+
+func (fh Firehose) Shutdown() {
+	if !fh.IsRunning() {
+		return
+	}
+	fh.shutdownMessage <- true
```

Comment:
There's a race condition here with two threads calling `Shutdown()`. Not sure about the context if it's a big deal, but it might be worth wrapping the checks around a `sync.Mutex` to make sure we don't deadlock sending to the `fh.shutdownMessage` channel.